server/app/services/scheduler/strategy/random.rb

```diff
@@ -13,22 +13,32 @@ def host_grace_period
         30.seconds
       end
 
-      ##
       # @param [GridService] grid_service
       # @param [Integer] instance_number
-      # @param [Array<HostNode>] nodes
+      # @param [Array<Scheduler::Node>] nodes
+      # @return [Scheduler::Node, NilClass]
       def find_node(grid_service, instance_number, nodes)
         if grid_service.stateless?
+          node = find_previous_or_new_node(grid_service, instance_number, nodes)
+          return node if node
           nodes.sample
         else
-          prev_instance = grid_service.grid_service_instances.find_by(
-            grid_service: grid_service, instance_number: instance_number
-          )
-          if prev_instance
-            prev_instance.host_node
-          else
-            nodes.sample
-          end
+          find_previous_or_new_node(grid_service, instance_number, nodes)
+        end
+      end
+
+      # @param [GridService] grid_service
+      # @param [Integer] instance_number
+      # @param [Array<Scheduler::Node>] nodes
+      # @return [Scheduler::Node, NilClass]
+      def find_previous_or_new_node(grid_service, instance_number, nodes)
+        prev_instance = grid_service.grid_service_instances.find_by(
+          grid_service: grid_service, instance_number: instance_number
+        )
+        if prev_instance
+          nodes.find{ |n| n.node == prev_instance.host_node }
```

Comment:
How does this behave for instances on removed nodes, which have `prev_instance.host_node => nil`?